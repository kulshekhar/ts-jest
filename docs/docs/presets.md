---
id: presets
title: Presets
---

### The 3 presets

`ts-jest` comes with 3 presets, covering most of the project's base configuration:

| Preset name | Description |
|---|---|
| `ts-jest/presets/default`<br/>or `ts-jest` | `ts-jest` will take care of `.ts` and `.tsx` files only, leaving JavaScript files as-is. |
| `ts-jest/presets/js-with-ts` | TypeScript and JavaScript files (`.ts`, `.tsx`, `.js` and `.jsx`) will be handled by `ts-jest`.<br/>You'll need to set `allowJs` to `true` in your `tsconfig.json` file. |
| `ts-jest/presets/js-with-babel` | TypeScript files will be handled by `ts-jest`, and JavaScript files will be handled by `babel-jest`. |

### Basic usage

In most cases, simply setting the `preset` key to the desired preset name in your Jest config should be enough to start using TypeScript with Jest (assuming you added `ts-jest` to your dev. dependencies of course):

```js
// jest.config.js
module.exports = {
  // [...]
  // Replace `ts-jest` with the preset you want to use
  // from the above list
  preset: 'ts-jest'
};
```

```js
// OR package.json
{
  // [...]
  "jest": {
    // Replace `ts-jest` with the preset you want to use
    // from the above list
    "preset": "ts-jest"
  }
}
```

**Note:** presets use `testMatch`, like Jest does in its defaults. If you want to use `testRegex` instead in your configuration, you MUST set `testMatch` to `null` or Jest will bail.

### Advanced

Any preset can also be used with other options.
If you're already using another preset, you might want only some specific settings from the chosen `ts-jest` preset.
In this case you'll need to use the JavaScript version of Jest config (comment/uncomment according to your use-case):

```js
// jest.config.js
const { defaults: tsjPreset } = require('ts-jest/presets');
// const { jsWithTs: tsjPreset } = require('ts-jest/presets');
// const { jsWithBabel: tsjPreset } = require('ts-jest/presets');

module.exports = {
  // [...]
  transform: {
    ...tsjPreset.transform,
    // [...]
  }
}
```
