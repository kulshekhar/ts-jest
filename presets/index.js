const { createJestPreset } = require('../dist/presets/create-jest-preset')

module.exports = {
  get defaults() {
    return createJestPreset()
  },
  get jsWithTs() {
    return createJestPreset({ allowJs: true })
  },
  get jsWithBabel() {
    return createJestPreset(
      { allowJs: false },
      {
        transform: {
          '^.+\\.jsx?$': 'babel-jest',
        },
      }
    )
  },
}
