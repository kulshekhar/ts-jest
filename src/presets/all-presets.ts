import {
  createDefaultEsmLegacyPreset,
  createDefaultEsmPreset,
  createDefaultLegacyPreset,
  createDefaultPreset,
  createJsWithBabelEsmLegacyPreset,
  createJsWithBabelEsmPreset,
  createJsWithBabelLegacyPreset,
  createJsWithBabelPreset,
  createJsWithTsEsmLegacyPreset,
  createJsWithTsEsmPreset,
  createJsWithTsLegacyPreset,
  createJsWithTsPreset,
} from './create-jest-preset'

const allPresets = {
  get defaults() {
    return createDefaultPreset()
  },
  get defaultsLegacy() {
    return createDefaultLegacyPreset()
  },
  get defaultsESM() {
    return createDefaultEsmPreset()
  },
  get defaultsESMLegacy() {
    return createDefaultEsmLegacyPreset()
  },
  get jsWithTs() {
    return createJsWithTsPreset()
  },
  get jsWithTsLegacy() {
    return createJsWithTsLegacyPreset()
  },
  get jsWithTsESM() {
    return createJsWithTsEsmPreset()
  },
  get jsWithTsESMLegacy() {
    return createJsWithTsEsmLegacyPreset()
  },
  get jsWithBabel() {
    return createJsWithBabelPreset()
  },
  get jsWithBabelLegacy() {
    return createJsWithBabelLegacyPreset()
  },
  get jsWithBabelESM() {
    return createJsWithBabelEsmPreset()
  },
  get jsWithBabelESMLegacy() {
    return createJsWithBabelEsmLegacyPreset()
  },
}

export default allPresets
