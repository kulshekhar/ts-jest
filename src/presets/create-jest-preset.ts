import type { Config } from '@jest/types'

import { rootLogger } from '../utils/logger'

const logger = rootLogger.child({ namespace: 'jest-preset' })

export type TsJestPresets = Pick<
  Config.InitialOptions,
  'extensionsToTreatAsEsm' | 'moduleFileExtensions' | 'transform' | 'testMatch'
>

export function createJestPreset(allowJs = false, extraOptions: Config.InitialOptions = {}): TsJestPresets {
  logger.debug({ allowJs }, 'creating jest presets', allowJs ? 'handling' : 'not handling', 'JavaScript files')

  const { extensionsToTreatAsEsm, moduleFileExtensions, testMatch } = extraOptions

  return {
    ...(extensionsToTreatAsEsm ? { extensionsToTreatAsEsm } : undefined),
    ...(moduleFileExtensions ? { moduleFileExtensions } : undefined),
    transform: {
      ...extraOptions.transform,
      [allowJs ? (extensionsToTreatAsEsm?.length ? '^.+\\.m?[tj]sx?$' : '^.+\\.[tj]sx?$') : '^.+\\.tsx?$']: 'ts-jest',
    },
    ...(testMatch ? { testMatch } : undefined),
  }
}
