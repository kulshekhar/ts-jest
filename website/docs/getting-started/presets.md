---
id: presets
title: Presets
---

### The presets

:::important

Starting from **v28.0.0**, `ts-jest` will gradually opt in adoption of `esbuild`/`swc` more to improve the performance. To make the transition smoothly, we introduce `legacy` presets as a fallback when the new codes don't work yet.

:::

:::caution

The list of `preset` below is now deprecated in favor of util functions. If one is using `preset` in Jest config, please run `npx ts-jest config:migrate` or look into [Advanced](#advanced) section below for alternative solutions.

:::

`ts-jest` comes with several presets, covering most of the project's base configuration:

| Preset name                                                           | Description                                                                                                                                                                                         |
| --------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `ts-jest/presets/default`<br/>or `ts-jest`                            | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **CommonJS** syntax, leaving JavaScript files (`.js`, `jsx`) as-is.                                                            |
| `ts-jest/presets/default-legacy`<br/>or `ts-jest/legacy` (**LEGACY**) | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **CommonJS** syntax, leaving JavaScript files (`.js`, `jsx`) as-is.                                                            |
| `ts-jest/presets/default-esm`<br/>                                    | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **ESM** syntax, leaving JavaScript files (`.js`, `jsx`) as-is.                                                                 |
| `ts-jest/presets/default-esm-legacy`<br/> (**LEGACY**)                | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **ESM** syntax, leaving JavaScript files (`.js`, `jsx`) as-is.                                                                 |
| `ts-jest/presets/js-with-ts`                                          | TypeScript and JavaScript files (`.ts`, `.tsx`, `.js`, `.jsx`) will be transformed by `ts-jest` to **CommonJS** syntax.<br/>You'll need to set `allowJs` to `true` in your `tsconfig.json` file.    |
| `ts-jest/presets/js-with-ts-legacy` (**LEGACY**)                      | TypeScript and JavaScript files (`.ts`, `.tsx`, `.js`, `.jsx`) will be transformed by `ts-jest` to **CommonJS** syntax.<br/>You'll need to set `allowJs` to `true` in your `tsconfig.json` file.    |
| `ts-jest/presets/js-with-ts-esm`                                      | TypeScript and JavaScript files (`.ts`, `.tsx`, `.js`, `.jsx`, `.mjs`) will be transformed by `ts-jest` to **ESM** syntax.<br/>You'll need to set `allowJs` to `true` in your `tsconfig.json` file. |
| `ts-jest/presets/js-with-ts-esm-legacy` (**LEGACY**)                  | TypeScript and JavaScript files (`.ts`, `.tsx`, `.js`, `.jsx`, `.mjs`) will be transformed by `ts-jest` to **ESM** syntax.<br/>You'll need to set `allowJs` to `true` in your `tsconfig.json` file. |
| `ts-jest/presets/js-with-babel`                                       | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **CommonJS** syntax, and JavaScript files (`.js`, `jsx`) will be transformed by `babel-jest`.                                  |
| `ts-jest/presets/js-with-babel-legacy` (**LEGACY**)                   | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **CommonJS** syntax, and JavaScript files (`.js`, `jsx`) will be transformed by `babel-jest`.                                  |
| `ts-jest/presets/js-with-babel-esm`                                   | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **ESM** syntax, and JavaScript files (`.js`, `jsx`, `.mjs`) will be transformed by `babel-jest`.                               |
| `ts-jest/presets/js-with-babel-esm-legacy` (**LEGACY**)               | TypeScript files (`.ts`, `.tsx`) will be transformed by `ts-jest` to **ESM** syntax, and JavaScript files (`.js`, `jsx`, `.mjs`) will be transformed by `babel-jest`.                               |

### Basic usage

In most cases, simply setting the `preset` key to the desired preset name in your Jest config should be enough to start
using TypeScript with Jest (assuming you added `ts-jest` to your `devDependencies` of course):

```js tab
// jest.config.js
const { createDefaultPreset } = require('ts-jest')

const defaultPreset = createDefaultPreset()

/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  // [...]
  // Replace `ts-jest` with the preset you want to use
  // from the above list
  ...defaultPreset,
}
```

```ts tab
// jest.config.ts
import { type JestConfigWithTsJest, createDefaultPreset } from 'ts-jest'

const defaultPreset = createDefaultPreset()

const jestConfig: JestConfigWithTsJest = {
  // [...]
  // Replace `ts-jest` with the preset you want to use
  // from the above list
  ...defaultPreset,
}

export default jestConfig
```

**Note:** presets use `testMatch`, like Jest does in its defaults. If you want to use `testRegex` instead in your configuration, you MUST set `testMatch` to `null` or Jest will bail.

### Advanced

There are several util functions to create and extend the existing presets:

- `createDefaultPreset`: for default preset
- `createDefaultLegacyPreset`: for default preset in legacy mode
- `createDefaultEsmPreset`: for default ESM preset
- `createDefaultEsmLegacyPreset`: for default ESM preset in legacy mode
- `createJsWithTsPreset`: for `js-with-ts` preset
- `createJsWithTsLegacyPreset`: for `js-with-ts` preset in legacy mode
- `createJsWithTsEsmPreset`: for `js-with-ts` ESM preset
- `createJsWithTsEsmLegacyPreset`: for `js-with-ts` ESM preset in legacy mode
- `createJsWithBabelPreset`: for `js-with-babel` preset
- `createJsWithBabelLegacyPreset`: for `js-with-babel` preset in legacy mode
- `createJsWithBabelEsmPreset`: for `js-with-babel` ESM preset
- `createJsWithBabelEsmLegacyPreset`: for `js-with-babel` ESM preset in legacy mode

Example:

```js tab
// jest.config.js
const { createDefaultPreset } = require('ts-jest')

/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  // [...]
  transform: {
    ...createDefaultPreset().transform,
    // [...]
  },
}
```

```ts tab
// jest.config.ts
import { createDefaultPreset, type JestConfigWithTsJest } from 'ts-jest'

const jestConfig: JestConfigWithTsJest = {
  // [...]
  transform: {
    ...createDefaultPreset().transform,
    // [...]
  },
}

export default jestConfig
```
