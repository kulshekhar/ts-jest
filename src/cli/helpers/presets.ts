import { TsJestPresets } from '../../config/create-jest-preset'

/** @internal */
export const enum JestPresetNames {
  default = 'ts-jest/presets/default',
  jsWithTs = 'ts-jest/presets/js-with-ts',
  jsWIthBabel = 'ts-jest/presets/js-with-babel',
}

/** @internal */
export interface TsJestPresetDescriptor {
  name: string
  fullName: string
  label: string
  jsVarName: string
  value: TsJestPresets
  isDefault: boolean
  jsImport(varName?: string): string
}

const definePreset = (fullName: string): TsJestPresetDescriptor => ({
  fullName,
  get name() {
    return this.isDefault ? 'ts-jest' : fullName
  },
  get label() {
    return fullName.split('/').pop()!
  },
  get jsVarName() {
    return this.isDefault
      ? 'defaults'
      : fullName
          .split('/')
          .pop()!
          // eslint-disable-next-line no-useless-escape
          .replace(/\-([a-z])/g, (_, l) => l.toUpperCase())
  },
  get value() {
    return require(`../../../${fullName.replace(/^ts-jest\//, '')}/jest-preset`)
  },
  jsImport(varName = 'tsjPreset') {
    return `const { ${this.jsVarName}: ${varName} } = require('ts-jest/presets')`
  },
  get isDefault() {
    return fullName === JestPresetNames.default
  },
})

/** @internal */
export const allPresets: Record<JestPresetNames, TsJestPresetDescriptor> = {} as any
/** @internal */
export const defaults = (allPresets[JestPresetNames.default] = definePreset(JestPresetNames.default))
/** @internal */
export const jsWithTs = (allPresets[JestPresetNames.jsWithTs] = definePreset(JestPresetNames.jsWithTs))
/** @internal */
export const jsWIthBabel = (allPresets[JestPresetNames.jsWIthBabel] = definePreset(JestPresetNames.jsWIthBabel))
