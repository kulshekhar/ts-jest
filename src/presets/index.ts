import { createJestPreset, TsJestPresets } from './create-jest-preset'

export = {
  get defaults(): TsJestPresets {
    return createJestPreset()
  },
  get jsWithTs(): TsJestPresets {
    return createJestPreset({ allowJs: true })
  },
  get jsWithBabel(): TsJestPresets {
    return createJestPreset(
      { allowJs: false },
      {
        transform: {
          '^.+\\.jsx?$': 'babel-jest',
        },
      },
    )
  },
}
