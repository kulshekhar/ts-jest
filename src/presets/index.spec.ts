import presets from './index'

describe('presets', () => {
  test.each([presets.defaults, presets.jsWithTs, presets.jsWithBabel])(
    'should return correct preset data',
    (presetData) => {
      expect(presetData).toMatchSnapshot()
    },
  )
})
