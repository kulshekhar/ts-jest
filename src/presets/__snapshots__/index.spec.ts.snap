// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presets should return correct preset data 1`] = `
Object {
  "transform": Object {
    "^.+\\\\.tsx?$": "ts-jest",
  },
}
`;

exports[`presets should return correct preset data 2`] = `
Object {
  "transform": Object {
    "^.+\\\\.[tj]sx?$": "ts-jest",
  },
}
`;

exports[`presets should return correct preset data 3`] = `
Object {
  "transform": Object {
    "^.+\\\\.jsx?$": "babel-jest",
    "^.+\\\\.tsx?$": "ts-jest",
  },
}
`;
